### PR TITLE
Replace config object with option pattern

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

File: correlation_id/config.go [hunk 1]
```diff
@@ -11,11 +11,37 @@ type Config struct {
...
-func NewConfig() *Config {
-	return &Config{
+func (c *Config) apply(options ...Option) *Config {
+	for _, option := range options {
+		option(c)
+	}
+	return c
+}
+
+// GetConfig return a new correlation configuration with all options applied
```
Comment (pimousse1099) [inline correlation_id/config.go]: return**s**

File: correlation_id/config.go [hunk 2]
```diff
@@ -11,11 +11,37 @@ type Config struct {
...
+func GetConfig(options ...Option) *Config {
```
Comment (pimousse1099) [inline correlation_id/config.go:22]: maybe declare this public func before the private `apply` func ?

File: correlation_id/config.go [hunk 3]
```diff
@@ -11,11 +11,37 @@ type Config struct {
...
+	config := &Config{
 		HeaderName: HeaderName,
 		IdGenerator: func(_ *http.Request) string {
 			return DefaultIdGenerator.Generate(10)
 		},
 	}
+	return config.apply(options...)
+}
+
+// Option was correlation middleware/tripperware configurable options
```
Comment (pimousse1099) [inline correlation_id/config.go]: > Option defines a correlationId middleware/tripperware configuration option

File: correlation_id/config.go [hunk 4]
```diff
@@ -11,11 +11,37 @@ type Config struct {
...
+type Option func(*Config)
+
+// WithHeaderName will configure HeaderName correlation options
```
Comment (pimousse1099) [inline correlation_id/config.go]: option~s~

File: correlation_id/config.go [hunk 5]
```diff
@@ -11,11 +11,37 @@ type Config struct {
...
+func WithHeaderName(headerName string) Option {
+	return func(config *Config) {
+		config.HeaderName = headerName
+	}
+}
+
+// WithIdGenerator will configure IdGenerator correlation options
```
Comment (pimousse1099) [inline correlation_id/config.go]: option~s~

File: middleware/correlation_id_test.go [hunk 6]
```diff
@@ -45,19 +45,15 @@ func TestCorrelationId(t *testing.T) {
...
-	config.IdGenerator = func(request *http.Request) string {
-		return "my-fixed-request-id"
-	}
-
 	// we recommend to use MiddlewareStack to simplify managing all wanted middlewares
 	// caution middleware order matters
 	stack := httpware.MiddlewareStack(
-		middleware.CorrelationId(config),
+		middleware.CorrelationId(
+			correlation_id.WithHeaderName("my-personal-header-name"),
```
Comment (pimousse1099) [inline middleware/correlation_id_test.go:52]: Nice one 👍

**Review by pimousse1099 — CHANGES_REQUESTED** [review]
nice one 👍 🎉 

> note: I commented one middleware about some typos but they are present in other the ones 😉

File: correlation_id/config.go (hunk 2 above)
Comment (instabledesign) [reply correlation_id/config.go:22]: i really prefer group object with his method and put the new func just after

**Review by pimousse1099 — APPROVED** [review]

**Pull request merged**